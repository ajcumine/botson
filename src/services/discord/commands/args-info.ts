import { Command } from './';

const argsInfo: Command = {
  argsRequired: true,
  description: 'Information about the arguments provided.',
  guildOnly: false,
  execute(message, args) {
    message.channel.send(
      `Arguments: ${args}\nArguments length: ${args.length}`,
    );
  },
};

export default argsInfo;
