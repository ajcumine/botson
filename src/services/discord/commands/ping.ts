import { Command } from './';

export const execute = message => {
  message.channel.send('Pong.');
};

const ping: Command = {
  execute,
  argsRequired: false,
  description: 'Ping!',
  guildOnly: false,
};

export default ping;
