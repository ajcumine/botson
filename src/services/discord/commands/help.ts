import { Message } from 'discord.js';
import { PREFIX } from '../';
import logger from '../../logger';
import commands, { Command } from './';

export const execute = async (message: Message, args: string[]) => {
  const data: string[] = [];
  if (!args.length) {
    data.push("Here's a list of all my commands:");
    data.push(
      Object.keys(commands)
        .map(commandKey => `\`${PREFIX}${commandKey}\``)
        .join(`, `),
    );
    data.push(
      `\nYou can send \`${PREFIX}help [command name]\` to get info on a specific command!`,
    );

    try {
      await message.author.send(data, { split: true });
      if (message.channel.type === 'dm') return;
      return message.reply("I've sent you a DM with all my commands!");
    } catch (error) {
      logger.error(`Could not send help DM to ${message.author.tag}.\n`, error);
      message.reply("it seems like I can't DM you! Do you have DMs disabled?");
    }
  }

  const commandName = args[0];
  const command = commands[commandName];

  if (!command) {
    return message.reply("that's not a valid command!");
  }

  data.push(`**Name:** ${commandName}`);
  data.push(`**Description:** ${command.description}`);
  if (command.usage) {
    data.push(`**Usage:** \`${PREFIX}${commandName} ${command.usage}\``);
  }

  return message.channel.send(data, { split: true });
};

const help: Command = {
  execute,
  description:
    'Provide a list of all the bot commands or specific info for a single command',
  argsRequired: false,
  guildOnly: false,
  usage: '[command name]',
};

export default help;
