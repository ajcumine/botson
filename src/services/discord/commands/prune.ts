import { Message } from 'discord.js';
import { Command } from './';

export const execute = (message: Message, args: string[]) => {
  const amount = parseInt(args[0], 10) + 1;
  if (isNaN(amount)) {
    return message.reply("that doesn't seem to be a valid number.");
  }

  if (amount <= 1 || amount > 100) {
    return message.reply('you need to input a number between 1 and 99.');
  }
  message.channel.bulkDelete(amount, true);
};

const prune: Command = {
  execute,
  description: 'Deletes a suppiled number of messages in the current channel',
  guildOnly: false,
  argsRequired: true,
  usage: '<number>',
};

export default prune;
