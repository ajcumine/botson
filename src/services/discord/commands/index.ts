import Discord from 'discord.js';
import argsInfo from './args-info';
import help from './help';
import ping from './ping';
import prune from './prune';
import role from './role';

export interface Command {
  description: string;
  argsRequired: boolean;
  usage?: string;
  guildOnly: boolean;
  execute: (message: Discord.Message, args: string[]) => void;
}

export interface Commands {
  [key: string]: Command;
}

const commands: Commands = {
  argsInfo,
  help,
  ping,
  prune,
  role,
};

export default commands;
