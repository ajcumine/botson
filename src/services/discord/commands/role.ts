import { Message } from 'discord.js';
import { Command } from '.';

export const execute = (message: Message, args: string[]) => {
  const taggedUser = message.mentions.users.first();
  const selectedRole = args.slice(1).join(' ');
  const guildRoles = message.guild.roles;
  if (guildRoles.find(gr => gr.name === selectedRole)) {
    message.channel.send(`You tried to set ${taggedUser} to ${selectedRole}`);
  } else {
    message.reply(
      `the role ${selectedRole} does not exist, available server roles are: ${guildRoles
        .map(gr => `${gr.name}`)
        .join(', ')}`,
    );
  }
};

const role: Command = {
  execute,
  argsRequired: true,
  description: 'Change a users role',
  usage: '<user> <role>',
  guildOnly: true,
};

export default role;
