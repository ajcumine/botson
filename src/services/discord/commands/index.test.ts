import commands from '.';

describe('commands', () => {
  it('is a list of command handlers', () => {
    expect(commands).toMatchSnapshot();
  });
});
