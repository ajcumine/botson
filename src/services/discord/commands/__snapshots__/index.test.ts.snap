// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands is a list of command handlers 1`] = `
Object {
  "help": Object {
    "argsRequired": false,
    "description": "Provide a list of all the bot commands or specific info for a single command",
    "execute": [Function],
    "guildOnly": false,
    "usage": "[command name]",
  },
  "ping": Object {
    "argsRequired": false,
    "description": "Ping!",
    "execute": [Function],
    "guildOnly": false,
  },
  "prune": Object {
    "argsRequired": true,
    "description": "Deletes a suppiled number of messages in the current channel",
    "execute": [Function],
    "guildOnly": false,
    "usage": "<number>",
  },
  "role": Object {
    "argsRequired": true,
    "description": "Change a users role",
    "execute": [Function],
    "guildOnly": true,
    "usage": "<user> <role>",
  },
}
`;
